### PR TITLE
【KernelGen】Add index_copy_ operator

### PR DESCRIPTION
**Pull request by Schopenhauer-loves-Hegel**
### PR Category
Operator

### Type of Change
New Feature

### Description
Add `index_copy_` operator implementation with Triton kernel.

- Implementation mode: `manual_kernel`
- Accuracy test: 24/24 passed

### Issue
N/A

### Progress

- [ ] Change is properly reviewed (1 reviewer required, 2 recommended).
- [ ] Change is responded to an issue.
- [x] Change is fully covered by a UT.

### Performance
**torch.float16**

| Shape | Torch Latency (ms) | Gems Latency (ms) | Speedup |
|-------|-------------------:|-------------------:|--------:|
| [64, 64] | 0.0150 | 0.5200 | 0.029 |
| [256, 256] | 0.0160 | 0.5170 | 0.031 |
| [1024, 1024] | 0.0230 | 0.5180 | 0.044 |
| [4096, 4096] | 0.2280 | 0.5690 | 0.401 |
| [1024, 65536] | 1.0030 | 1.3320 | 0.753 |
| [10000, 256] | 0.0280 | 0.5120 | 0.055 |
| [10000, 65536] | 9.6590 | 11.7830 | 0.820 |

**torch.float32**

| Shape | Torch Latency (ms) | Gems Latency (ms) | Speedup |
|-------|-------------------:|-------------------:|--------:|
| [64, 64] | 0.0130 | 0.5120 | 0.025 |
| [256, 256] | 0.0140 | 0.5110 | 0.027 |
| [1024, 1024] | 0.0280 | 0.5650 | 0.049 |
| [4096, 4096] | 0.3140 | 0.5730 | 0.548 |
| [1024, 65536] | 1.3980 | 1.3760 | 1.016 |
| [10000, 256] | 0.0370 | 0.5160 | 0.072 |
| [10000, 65536] | 13.5900 | 12.2410 | 1.110 |

**Overall: median speedup = 0.064x, mean speedup = 0.356x** (14 data points)

---
_Generated by auto_gen tool with Claude Code_

### CODE REVIEW

**Comment by factnn** [comment]
## index_copy_ Benchmark Speedup Results

### float16
| Size | Torch (ms) | Gems (ms) | Speedup |
|------|-----------|----------|---------|
| [64, 64] | 0.016 | 0.717 | 0.022 |
| [256, 256] | 0.016 | 0.702 | 0.023 |
| [1024, 1024] | 0.027 | 0.731 | 0.036 |
| [4096, 4096] | 0.228 | 0.839 | 0.272 |
| [1024, 65536] | 1.005 | 1.383 | 0.726 |
| [10000, 65536] | 9.645 | 11.881 | 0.812 |

### float32
| Size | Torch (ms) | Gems (ms) | Speedup |
|------|-----------|----------|---------|
| [64, 64] | 0.013 | 0.722 | 0.018 |
| [256, 256] | 0.014 | 0.756 | 0.018 |
| [1024, 1024] | 0.028 | 0.737 | 0.038 |
| [4096, 4096] | 0.315 | 0.828 | 0.381 |
| [1024, 65536] | 1.393 | 1.460 | 0.954 |
| [10000, 65536] | 13.617 | 12.335 | 1.104 |

Note: Significant performance regression on small tensors due to kernel launch overhead. Large tensors (float32) show comparable or better performance.

File: src/flag_gems/ops/index_copy_.py [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+_index_copy_func = IndexCopyFunction()
+
+
+_FALLBACK_KEYSET = torch._C.DispatchKeySet(
+    torch._C.DispatchKey.CompositeExplicitAutograd
+)
+
+
+def index_copy(inp, dim, index, src):
+    logger.debug("GEMS INDEX COPY")
```
Comment (tengqm) [inline src/flag_gems/ops/index_copy_.py]:
```suggestion
    logger.debug("GEMS INDEX_COPY")
```

File: src/flag_gems/ops/index_copy_.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+        src_shape_dim,
+        delta,
+        N,
+        inp.numel(),
+    )
+    return out
+
+
+def index_copy_(inp, dim, index, src):
+    logger.debug("GEMS INDEX COPY_")
```
Comment (tengqm) [inline src/flag_gems/ops/index_copy_.py]:
```suggestion
    logger.debug("GEMS INDEX_COPY_")
```

**Review by tengqm — COMMENTED** [review]
overall looks good

File: benchmark/test_index_copy_perf.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    index_max = src_shape[dim]
+    index_len = index_max // 2 if index_max >= 2 else 1
+    index = torch.randperm(index_len, device=device)
+    src_shape[dim] = index_len
+    src = torch.randn(src_shape, dtype=dtype, device=device)
+    yield inp, dim, index, src
+
+
+@pytest.mark.index_copy
+def test_index_copy_perf():
```
Comment (tengqm) [inline benchmark/test_index_copy_perf.py]:
```suggestion
def test_index_copy():
```

File: benchmark/test_index_copy_perf.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    bench = TensorSelectBenchmark(
+        op_name="index_copy",
+        torch_op=torch.index_copy,
+        input_fn=index_copy_input_fn,
+        dtypes=[torch.float16, torch.float32],
```
Comment (tengqm) [inline benchmark/test_index_copy_perf.py]: why skip bf16?

File: benchmark/test_index_copy_perf.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    bench.run()
+
+
+@pytest.mark.index_copy_
+def test_index_copy__perf():
+    bench = TensorSelectBenchmark(
+        op_name="index_copy_",
+        torch_op=torch.Tensor.index_copy_,
+        input_fn=index_copy_input_fn,
+        dtypes=[torch.float16, torch.float32],
```
Comment (tengqm) [inline benchmark/test_index_copy_perf.py]: same question here

File: benchmark/test_index_copy_perf.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
+import pytest
+import torch
+
+from benchmark.performance_utils import GenericBenchmark2DOnly
+from flag_gems.utils import shape_utils
+
+
+class TensorSelectBenchmark(GenericBenchmark2DOnly):
+    pass
```
Comment (tengqm) [inline benchmark/test_index_copy_perf.py]: This additional layer makes no sense.

File: benchmark/test_index_copy_perf.py (hunk 5 above)
Comment (tengqm) [reply benchmark/test_index_copy_perf.py]: shall we specify `inplace=True`?

File: PR_status.md [hunk 7]
```diff
@@ -0,0 +1,45 @@
+# Feb 27 KernelGen PR Status
```
Comment (tengqm) [inline PR_status.md]: Remove this file.

**Comment by factnn** [comment]
## UT Results

Branch rebased to upstream/master and tests passed:

```
python -m pytest tests/ -m index_copy_ --ref cpu -x -q
12 passed
```

**Comment by bin913** [comment]
Could you please introduce adopting the code generation approach to generate kernel code for index_copy, and what are the advantages of doing so?

File: src/flag_gems/ops/index_copy_.py [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+    assert index.numel() == src.size(
+        dim
+    ), "The dimth dimension of source must have the same size as the length of index"
+    assert (
+        inp.ndim == src.ndim
+    ), "Self and source should have the same number of dimensions"
+    assert (
+        ((inp.size(i) == src.size(i)) or i == dim) for i in range(0, inp.ndim)
+    ), "src.size(d) == self.size(d) for all dimensions d != dim"
+
```
Comment (bin913) [inline src/flag_gems/ops/index_copy_.py:278]: The assertion is always true since a generator expression is truthy.

**Comment by factnn** [comment]
> Could you please introduce adopting the code generation approach to generate kernel code for index_copy, and what are the advantages of doing so?

Since I didn't implement this operator myself, my only responsibility is to submit this PR. I don’t actually have in-depth knowledge of the operator’s internal details, so I can only provide a response generated by Claude Code.

The index_copy_ operator needs to handle tensors of arbitrary rank. With a standard @triton.jit kernel, the function signature is fixed — you can't dynamically vary the number of stride and shape parameters based on rank at compile time.
  The code generation approach solves this by generating a specialized kernel for each rank at runtime, where the exact number of stride/shape arguments is known. This avoids dynamic indexing loops inside the kernel and lets Triton compile a fully unrolled, rank-specific kernel — giving better performance than a generic fallback.
  This pattern is consistent with how other rank-dependent operators in FlagGems (e.g. gather, scatter) are implemented.

File: src/flag_gems/ops/index_copy_.py [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+class IndexCopyFunction:
+    def __init__(self):
+        self.pid = os.getpid()
+        self.overloads: Mapping[str, Callable] = {}
+
+    def __call__(self, *args, **kwargs):
+        key = f"{self.arg_key(*args)}"
+        if key in self.overloads:
+            overload = self.overloads[key]
```
Comment (tengqm) [inline src/flag_gems/ops/index_copy_.py]: If you return here with `return overload(*args, **kwargs)` here, you don't need the `else` on line 176.

File: src/flag_gems/ops/index_copy_.py [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+            # load
+            spec = importlib.util.spec_from_file_location(
+                f"_gen_module_rank_{key}_pid_{self.pid}",
+                f.name,
+            )
+
+            m = importlib.util.module_from_spec(spec)
+            spec.loader.exec_module(m)
+            overload = getattr(m, "_index_copy_wrapper")
+            self.overloads[key] = overload
```
Comment (tengqm) [inline src/flag_gems/ops/index_copy_.py:200]: Please consider wrap these logic into a `try...except` block.

Comment (factnn) [reply src/flag_gems/ops/index_copy_.py:200]: done

File: src/flag_gems/ops/index_copy_.py (hunk 9 above)
Comment (factnn) [reply src/flag_gems/ops/index_copy_.py]: done

File: benchmark/test_index_copy.py [hunk 11]
```diff
@@ -0,0 +1,53 @@
+import pytest
```
Comment (tengqm) [inline benchmark/test_index_copy.py:1]: remove `_perf` from the file name.

File: benchmark/test_index_copy_perf.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+    index_max = src_shape[dim]
+    index_len = index_max // 2 if index_max >= 2 else 1
+    index = torch.randperm(index_len, device=device)
+    src_shape[dim] = index_len
+    src = torch.randn(src_shape, dtype=dtype, device=device)
+    yield inp, dim, index, src
+
+
+def _inplace_input_fn(shape, dtype, device):
+    yield from _tensor_input_fn(shape, dtype, device)
```
Comment (tengqm) [inline benchmark/test_index_copy_perf.py]: Why this wrapper?

File: conf/operators.yaml
Comment (tengqm) [inline conf/operators.yaml:5219]: Why this change?

File: src/flag_gems/ops/index_copy_.py (hunk 9 above)
Comment (tengqm) [reply src/flag_gems/ops/index_copy_.py]: No. You didn't return, still.

File: src/flag_gems/ops/index_copy_.py (hunk 10 above)
Comment (tengqm) [reply src/flag_gems/ops/index_copy_.py:200]: You ignored my review, again.

File: src/flag_gems/ops/index_copy_.py [hunk 13]
```diff
@@ -0,0 +1,298 @@
...
+        if key in self.overloads:
+            return self.overloads[key](*args, **kwargs)
+
+        code = IndentedBuffer()
+        code = generate_code(
+            args,
+            "_index_copy_wrapper",
+            "_index_copy_jit_function",
+            code,
+        )
```
Comment (tengqm) [inline src/flag_gems/ops/index_copy_.py:183]: just out of curiosity ... why use generated Python code rather than programming the function in Python directly?

Comment (factnn) [reply src/flag_gems/ops/index_copy_.py:183]: The kernel signature requires a fixed number of stride arguments at compile time (one per dimension: src_stride_0, src_stride_1, ...). Since the tensor rank is only known at runtime, we use code generation to produce a specialized kernel for each rank encountered. This is the same pattern used by other rank-dependent operators in FlagGems.

Comment (tengqm) [reply src/flag_gems/ops/index_copy_.py:183]: Got it. Thank you.

**Review by bin913 — APPROVED** [review]
